File: DairoNPC-Android/app/src/main/java/cn/dairo/npc/repository/NpcRepository.kt
package cn.dairo.npc.repository

import android.content.Context
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import cn.dairo.npc.bean.NpcSetting
import cn.dairo.npc.extension.npcDataStore
import com.google.gson.Gson
import kotlinx.coroutines.flow.first

class NpcRepository(private val context: Context) {

    /**
     * 配置存储key
     */
    private val settingKey = stringPreferencesKey("setting")

    /**
     * 标记是否是打开状态的
     */
    private val isOpenedKey = booleanPreferencesKey("isOpened")

    /**
     * 是否已经配置标记存储key
     */
    private val isSetKey = booleanPreferencesKey("isSet")

    /**
     * 从缓存加载
     */
    suspend fun loadSetting(): NpcSetting {
        val pref = this.context.npcDataStore.data.first()
        val json = pref[settingKey]?:return NpcSetting()
        return Gson().fromJson(json, NpcSetting::class.java)
    }

    /**
     * 保存配置
     */
    suspend fun saveSetting(bean: NpcSetting){
        this.context.npcDataStore.edit {
            it[settingKey] = Gson().toJson(bean)
        }
    }

    /**
     * 获取标记是否是打开状态
     */
    suspend fun isOpened(): Boolean {
        val pref = this.context.npcDataStore.data.first()
        return pref[isOpenedKey] ?: false
    }

    /**
     * 设置标记是否是打开状态
     */
    suspend fun setOpened(flag:Boolean) {
        this.context.npcDataStore.edit {
            it[isOpenedKey] = flag
        }
    }

    /**
     * 获取是否已经设置完成
     */
    suspend fun isSet(): Boolean {
        val pref = this.context.npcDataStore.data.first()
        return pref[isSetKey] ?: false
    }

    /**
     * 获取是否已经设置完成
     */
    suspend fun saveSet(flag:Boolean) {
        this.context.npcDataStore.edit {
            it[isSetKey] = flag
        }
    }

}

File: DairoNPC-IOS/DairoNPC/Repository/NpcRepository.swift
import Foundation
class NpcRepository {
    
    /**
     * 配置存储key
     */
    private static let settingKey = "setting"
    
    /**
     * 标记是否是打开状态的
     */
    private static let isOpenedKey = "isOpened"
    
    /**
     * 是否已经配置标记存储key
     */
    private static let isSetKey = "isSet"
    
    /**
     * 从缓存加载
     */
    static func loadSetting() -> NpcSetting {
        if let jsonData = UserDefaults.standard.data(forKey: settingKey){
            if let npcSetting = try? JSONDecoder().decode(NpcSetting.self, from: jsonData){
                return npcSetting
            }
        }
        return NpcSetting()
    }
    
    /**
     * 保存配置
     */
    static func saveSetting(_ bean: NpcSetting){
        let encoder = JSONEncoder()
        let data = try! encoder.encode(bean)
        UserDefaults.standard.set(data, forKey: settingKey)
    }
    
    /**
     * 获取标记是否是打开状态
     */
    static func isOpened() -> Bool {
        return UserDefaults.standard.bool(forKey: isOpenedKey)
    }
    
    /**
     * 设置标记是否是打开状态
     */
    static func setOpened(_ flag: Bool) {
        UserDefaults.standard.set(flag, forKey:isOpenedKey)
    }
    
    /**
     * 获取是否已经设置完成
     */
    static func isSet() -> Bool {
        return UserDefaults.standard.bool(forKey: isSetKey)
    }
    
    /**
     * 获取是否已经设置完成
     */
    static func saveSet(_ flag: Bool) {
        UserDefaults.standard.set(flag, forKey: isSetKey)
    }
}
